include/trieste/rewrite.h

```diff
@@ -1035,6 +1041,112 @@ namespace trieste
     };
   }
 
+  detail::PatternEffect<Node> Dbg(Token, detail::PatternEffect<Node>, bool);
+
+  namespace detail
+  {
+    class DebugMatch : public PatternDef
+    {
+    private:
+      Token start;
+      Pattern pattern;
+      bool dbg_on_success;
+
+      std::vector<std::string> notes;
+      std::set<Token> captures;
+      int success_count;
+
+      friend PatternEffect<Node> trieste::Dbg(Token, PatternEffect<Node>, bool);
+
+    public:
+      DebugMatch(Token start_, Pattern pattern_, bool dbg_on_success_ = false)
+      : start(start_), pattern(pattern_), dbg_on_success(dbg_on_success_)
+      {}
+
+      PatternPtr clone() const& override
+      {
+        throw std::runtime_error("The DebugMatch pattern can't be cloned");
+      };
+
+      bool match(NodeIt& it, const Node& parent, Match& match) const& override
+      {
```

Comment:
I'm unsure on what this will produce in the failure case for a disjunction?  Will it only report the failure of the final disjunct?

Reply:
Yes, this is sadly a limitation of this non-intrusive approach.